### PR TITLE
Add simple neuromorphic energy estimation

### PR DESCRIPTION
**Pull request by AllenYolk**
## Summary

- replace the compute-only estimator with `SimpleEnergy`, combining runtime MAC/AC counts with logical neuromorphic memory accesses
- add an independent neuromorphic memory counter for runtime parameter and persistent neuron-state traffic
- align Lemaire memory accounting with per-layer SRAM capacities and exact event fanout geometry

## Validation

- `uv run pytest -q test/activation_based/test_ac.py test/activation_based/test_mac.py test/activation_based/test_op_counter_*.py test/test_logging_policy.py` (`182 passed`)
- g2: `pytest -q test/activation_based` (`1731 passed, 57 skipped`)
- `uv run ruff check ...`
- `uv run python tools/generate_changelog_rst.py --check`
- `uv run sphinx-build -M html docs/source docs/build`
- `git diff --check`

## Summary by CodeRabbit

- **New Features**
  - Added runtime energy estimation for compute, parameter, bias, and neuron-state memory access.
  - Added neuromorphic memory-access reporting for binary and non-binary workloads.
  - Added FP32, FP16, and INT8 cost presets, plus configurable SNN FIFO capacity.

- **Breaking Changes**
  - Replaced legacy `ComputeEnergy*` APIs with `SimpleEnergy*` equivalents.

- **Bug Fixes**
  - Improved memory accounting and handling of unsupported transposed convolutions.

- **Documentation**
  - Updated English and Chinese tutorials, API references, and changelogs.

### CODE REVIEW

File: test/activation_based/test_op_counter_snn_energy.py [hunk 1]
```diff
@@ -260,8 +330,32 @@ def test_lemaire_energy_linear_inout_uses_runtime_dtype_bytes():
...
+    assert report.counts["write_out_bytes"] == 3 * 4
+    assert (
+        report.counts["read_potential_bytes"]
+        == (event_fanout + base_potential_updates) * 4
+    )
+    assert (
+        report.counts["write_potential_bytes"]
+        == (event_fanout + base_potential_updates) * 4
     )
+    assert report.buffer_sizes_bytes["inout_buffer_bytes"] == 1000 * 4
```
[the remark this answers is not in the document]
Comment (AllenYolk) [reply test/activation_based/test_op_counter_snn_energy.py:370]: No change. The literal 1000 intentionally protects the documented default FIFO-capacity contract. Reading the expected value from LemaireEnergyConfig() would make the test pass after an accidental default change and weaken the regression.

File: spikingjelly/activation_based/op_counter/neuromorphic_memory_access.py [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+                if module.bias is None
+                else int(output.numel()) * int(module.bias.element_size())
+            ),
+        )
+
+    def _neuron_hook(
+        self, module: BaseNode, inputs: tuple[torch.Tensor, ...], output: Any
+    ) -> None:
+        del output
+        time_steps = int(inputs[0].shape[0]) if module.step_mode == "m" else 1
```
[the remark this answers is not in the document]
Comment (AllenYolk) [reply spikingjelly/activation_based/op_counter/neuromorphic_memory_access.py:161]: No change. Multi-step BaseNode input layout [T, N, *] is the established SpikingJelly step-mode contract. A transpose adapter that moves T elsewhere must restore that contract before invoking the node; inferring an arbitrary time axis here would add ambiguity and silently guess.

File: test/activation_based/test_op_counter_snn_energy.py [hunk 3]
```diff
@@ -260,8 +330,32 @@ def test_lemaire_energy_linear_inout_uses_runtime_dtype_bytes():
     assert report.counts["read_in_bytes"] == x.numel() * x.element_size()
     assert report.counts["write_out_bytes"] == (3 * 4) * x.element_size()
     assert report.counts["read_params_bytes"] == (
-        model.weight.numel() * model.weight.element_size()
+        3 * model.weight.numel() * model.weight.element_size()
+    )
```
[the remark this answers is not in the document]
Comment (AllenYolk) [reply test/activation_based/test_op_counter_snn_energy.py:346]: No change. For every nn.Linear shape, batch * weight.numel() equals out.numel() * in_features; the literal 3 deliberately asserts the batch-3 test case. Replacing it with x.shape[0] would make the expected value less explicit without changing coverage.

File: spikingjelly/activation_based/op_counter/neuromorphic_memory_access.py [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+            torch.ones_like(module.weight, dtype=torch.float64),
+            None,
+            module.stride,
+            padding,
+            module.dilation,
+            False,
+            tuple(0 for _ in module.stride),
+            module.groups,
+        )
+    return int(out.sum().item())
```
[the remark this answers is not in the document]
Comment (AllenYolk) [reply spikingjelly/activation_based/op_counter/neuromorphic_memory_access.py:54]: Valid and fixed in 4da8ce90. String padding now uses the module's explicit reversed padding through F.pad and passes zero padding to aten.convolution. Both same and valid have regression cases.

File: spikingjelly/activation_based/op_counter/analytical_energy/core.py [hunk 5]
```diff
@@ -180,151 +194,185 @@ class LemaireEnergyReport:
...
             )
-            if not torch.is_tensor(x) or not torch.is_tensor(out):
-                self._track_bytes(
-                    dense_bytes_tree(inputs),
-                    dense_bytes_tree(output),
-                    sum(
-                        dense_bytes(param) for param in module.parameters(recurse=False)
-                    ),
+            is_spike_input = is_binary_tensor(x)
+            active_inputs = int(x.count_nonzero().item()) if is_spike_input else 0
```
[the remark this answers is not in the document]
Comment (AllenYolk) [reply spikingjelly/activation_based/op_counter/analytical_energy/core.py]: Valid, with a broader root cause fixed in 4da8ce90. Guarding tracker-only tensor operations prevents measurement work from entering counters, but that alone does not make the composed strict DispatchCounterMode usable: supported Linear already fails on aten.t because the specialized counters intentionally have different rule sets. The composite mode is now non-strict while config.strict remains enforced by the backend, Lemaire-formula, and neuron-state owners. A strict supported-Linear regression test was added.

File: spikingjelly/activation_based/op_counter/neuromorphic_memory_access.py [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+        }
+
+    def get_total(self) -> int:
+        r"""
+        返回全局访存字节总数。Return the global total memory-access bytes.
+
+        :return: 全局访存字节总数 / Global total bytes
+        :rtype: int
+        """
+        return sum(self._records["Global"].values())
```
[the remark this answers is not in the document]
Comment (AllenYolk) [reply spikingjelly/activation_based/op_counter/neuromorphic_memory_access.py]: Valid and fixed in 4da8ce90. get_total() now reads Global with dict.get(), so querying an empty counter no longer creates a scope. A regression test covers the no-data state.

File: test/activation_based/test_op_counter_snn_energy.py [hunk 7]
```diff
@@ -188,6 +230,34 @@ def test_lemaire_energy_cost_config_converts_access_costs_to_pj_per_byte():
...
+    report = op_counter.estimate_lemaire_energy(
+        model,
+        torch.full((1, 2), 0.5),
+        config=op_counter.LemaireEnergyConfig(cost_config=cost),
+    )
+
+    # Layer capacities are 16 B and 64 B; their read traffic is also 16 B and 64 B.
+    assert report.counts["read_params_bytes"] == 80
+    assert report.breakdown_pj["params_pj"] == pytest.approx(16 * 1 + 64 * 4)
+    assert report.breakdown_pj["params_pj"] != pytest.approx(80 * 4)
```
[the remark this answers is not in the document]
Comment (AllenYolk) [reply test/activation_based/test_op_counter_snn_energy.py:270]: No change. The existing comment already states that both layer capacities and read traffic are 16 B and 64 B. This fp32, batch-1 setup is deliberate because the test isolates per-layer SRAM pricing; expanding the comment would repeat the fixture rather than protect additional behavior.

File: CHANGELOG.md [hunk 8]
```diff
@@ -52,6 +59,16 @@ Module: `spikingjelly.activation_based.model`.
 
 ### Bug Fixes
 
+#### Operation Counters
```
[the remark this answers is not in the document]
Comment (AllenYolk) [reply CHANGELOG.md]: Valid and fixed in 4da8ce90. The Bug Fixes and Breaking Changes headings are now distinct, and the generated RST changelog was refreshed.
